File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/components/UseSmileIDSampleKeyValueEditRow.kt
package com.usesmileid.sampleapps.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.unit.sp
import com.smileid.designsystem.SmileDimens
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/** A label and a value that edits in place with a caret, rather than pushing a form. */
@OptIn(ExperimentalLayoutApi::class)
@Composable
fun UseSmileIDSampleKeyValueEditRow(
    label: String,
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    placeholder: String = "",
    required: Boolean = false,
    enabled: Boolean = true,
    keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
    testId: String? = null,
) {
    val colors = UseSmileIDSampleTheme.colors
    FlowRow(
        modifier = modifier
            .fillMaxWidth()
            .background(colors.surface)
            .defaultMinSize(minHeight = SmileDimens.sizeControlMd)
            .padding(horizontal = SmileDimens.spacingMd, vertical = SmileDimens.spacingSm),
        // The value sits at the right edge and drops below the label at 2x, matching DataFieldRow.
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalArrangement = Arrangement.spacedBy(SmileDimens.spacingXxs),
        itemVerticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            // The design puts the field name in the title colour; only the placeholder is muted.
            text = if (required) "$label *" else label,
            style = UseSmileIDSampleTheme.type.textStyleBody.copy(fontSize = ROW_TEXT_SIZE),
            color = colors.textTitle,
        )
        BasicTextField(
            value = value,
            onValueChange = onValueChange,
            enabled = enabled,
            singleLine = true,
            keyboardOptions = keyboardOptions,
            textStyle = UseSmileIDSampleTheme.type.textStyleBody.copy(
                fontSize = ROW_TEXT_SIZE,
                // Muted when disabled, so a row that cannot be edited does not look editable.
                color = if (enabled) colors.textTitle else colors.textMuted,
            ),
            cursorBrush = SolidColor(colors.primary),
            modifier = Modifier.tagged(testId),
            decorationBox = { field ->
                if (value.isEmpty()) {
                    Text(
                        text = placeholder,
                        style = UseSmileIDSampleTheme.type.textStyleBody.copy(fontSize = ROW_TEXT_SIZE),
                        color = colors.textMuted,
                    )
                }
                field()
            },
        )
    }
}

/** 13.5 in the design for the field name, its value and its placeholder alike. */
private val ROW_TEXT_SIZE = 13.5.sp

File: flutter/sample_ui/lib/src/components/use_smileid_sample_key_value_edit_row.dart
import 'package:flutter/material.dart';

import '../theme/use_smileid_sample_colors.dart';
import '../theme/use_smileid_sample_theme.dart';
import '../theme/use_smileid_sample_typography.dart';
import '../tokens/smile_tokens.dart';

/// A label and a value that edits in place with a caret, rather than pushing a form.
///
/// The field NAME is title-coloured and only the PLACEHOLDER is muted; Android had that inverted.
class UseSmileIDSampleKeyValueEditRow extends StatefulWidget {
  /// [required] appends an asterisk to the label, which is how the design marks a required field.
  const UseSmileIDSampleKeyValueEditRow({
    required this.label,
    required this.value,
    required this.onChanged,
    this.placeholder = '',
    this.required = false,
    this.enabled = true,
    this.keyboardType,
    this.testId,
    super.key,
  });

  /// The field name.
  final String label;

  /// The current value.
  final String value;

  /// Called on every edit.
  final ValueChanged<String> onChanged;

  /// Shown while the value is empty, and the only muted run in the row.
  final String placeholder;

  /// Whether the field is required.
  final bool required;

  /// Whether the row accepts input.
  final bool enabled;

  /// The keyboard to raise.
  final TextInputType? keyboardType;

  /// The `sample_*` id the screen supplies.
  final String? testId;

  @override
  State<UseSmileIDSampleKeyValueEditRow> createState() =>
      _UseSmileIDSampleKeyValueEditRowState();
}

class _UseSmileIDSampleKeyValueEditRowState
    extends State<UseSmileIDSampleKeyValueEditRow> {
  late final TextEditingController _controller = TextEditingController(
    text: widget.value,
  );

  @override
  void didUpdateWidget(UseSmileIDSampleKeyValueEditRow oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.value != _controller.text) {
      _controller.value = _controller.value.copyWith(
        text: widget.value,
        selection: TextSelection.collapsed(offset: widget.value.length),
        composing: TextRange.empty,
      );
    }
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    final TextStyle rowStyle = UseSmileIDSampleType.textStyleSubtitle.copyWith(
      fontSize: _rowTextSize,
    );
    return ColoredBox(
      color: colors.surface,
      child: ConstrainedBox(
        constraints: const BoxConstraints(minHeight: SmileDimens.sizeControlMd),
        child: Padding(
          padding: const EdgeInsets.symmetric(
            horizontal: _rowPaddingX,
            vertical: _rowPaddingY,
          ),
          // The value sits at the right edge and drops below the label at 2x, matching DataFieldRow.
          child: Wrap(
            alignment: WrapAlignment.spaceBetween,
            runSpacing: SmileDimens.spacingXxs,
            crossAxisAlignment: WrapCrossAlignment.center,
            children: <Widget>[
              Text(
                widget.required ? '${widget.label} *' : widget.label,
                style: rowStyle.copyWith(color: colors.textTitle),
              ),
              Semantics(
                identifier: widget.testId,
                textField: true,
                child: IntrinsicWidth(
                  child: TextField(
                    controller: _controller,
                    enabled: widget.enabled,
                    onChanged: widget.onChanged,
                    keyboardType: widget.keyboardType,
                    maxLines: 1,
                    textAlign: TextAlign.end,
                    cursorColor: colors.primary,
                    // Muted when disabled, so a row that cannot be edited does not look editable.
                    style: rowStyle.copyWith(
                      color: widget.enabled
                          ? colors.textTitle
                          : colors.textMuted,
                    ),
                    decoration: InputDecoration.collapsed(
                      hintText: widget.placeholder,
                      hintStyle: rowStyle.copyWith(color: colors.textMuted),
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

/// Label, value and placeholder alike — NOT text-style.body, which is weight 400.
const double _rowTextSize = 13.5;

/// The design's own padding; no scale token carries 15 or 14.
const double _rowPaddingX = 15;
const double _rowPaddingY = 14;
